hedera-node/src/main/java/com/hedera/services/txns/token/TokenUpdateTransitionLogic.java

```diff
@@ -65,125 +67,195 @@
 		}
+	}
 
-		Optional<AccountID> replacedTreasury = Optional.empty();
-		if (op.hasTreasury()) {
-			var newTreasury = op.getTreasury();
-			if (ledger.isDetached(newTreasury)) {
-				txnCtx.setStatus(ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
-				return;
-			}
-			if (!store.associationExists(newTreasury, id)) {
-				txnCtx.setStatus(INVALID_TREASURY_ACCOUNT_FOR_TOKEN);
-				return;
-			}
-			var existingTreasury = token.treasury().toGrpcAccountId();
-			if (!allowChangedTreasuryToOwnNfts && token.tokenType() == NON_FUNGIBLE_UNIQUE) {
-				var existingTreasuryBalance = ledger.getTokenBalance(existingTreasury, id);
-				if (existingTreasuryBalance > 0L) {
-					abortWith(CURRENT_TREASURY_STILL_OWNS_NFTS);
-					return;
-				}
-			}
-			if (!newTreasury.equals(existingTreasury)) {
-				if (ledger.isDetached(existingTreasury)) {
-					txnCtx.setStatus(ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
-					return;
-				}
-				outcome = prepNewTreasury(id, token, newTreasury);
-				if (outcome != OK) {
-					abortWith(outcome);
-					return;
-				}
-				replacedTreasury = Optional.of(token.treasury().toGrpcAccountId());
-			}
+	private void prepareNewTreasury(final TokenRelationship newTreasuryRel, final Token token) {
+		if (token.hasFreezeKey()) {
+			newTreasuryRel.changeFrozenState(false);
+		}
+		if (token.hasKycKey()) {
+			newTreasuryRel.changeKycState(true);
 		}
 
-		outcome = store.update(op, txnCtx.consensusTime().getEpochSecond());
-		if (outcome == OK && replacedTreasury.isPresent()) {
-			final var oldTreasury = replacedTreasury.get();
-			long replacedTreasuryBalance = ledger.getTokenBalance(oldTreasury, id);
-			if (replacedTreasuryBalance > 0) {
-				if (token.tokenType().equals(TokenType.FUNGIBLE_COMMON)) {
-					outcome = ledger.doTokenTransfer(
-							id,
-							oldTreasury,
-							op.getTreasury(),
-							replacedTreasuryBalance);
+		tokenStore.persistTokenRelationships(List.of(newTreasuryRel));
+	}
+
+	private void persistTransferOfTreasuryBalanceIfAppropriate(
+			final Token token,
+			final Account currentTreasury,
+			final Account newTreasury) {
+
+		final var transferIsAppropriate = newTreasury != null;
+
+		if (transferIsAppropriate) {
+			final var currentTreasuryRel = tokenStore.loadTokenRelationship(token, currentTreasury);
+			final var treasuryBalanceToBeReplaced = currentTreasuryRel.getBalance();
+
+			final var treasuryBalanceIsEligibleForTransfer = treasuryBalanceToBeReplaced > 0L;
+			final var isFungibleCommon = token.getType().equals(TokenType.FUNGIBLE_COMMON);
+
+			if (treasuryBalanceIsEligibleForTransfer) {
+				if (isFungibleCommon) {
+					doFungibleTransfer(token, currentTreasury, newTreasury, treasuryBalanceToBeReplaced);
 				} else {
-					outcome = store.changeOwnerWildCard(
-							new NftId(id.getShardNum(), id.getRealmNum(), id.getTokenNum(), -1), oldTreasury, op.getTreasury());
+					doNonFungibleTransfer(token, currentTreasury, newTreasury, tokenStore);
 				}
 			}
```

Comment:
You should use the `doZeroSum` methods for various reasons, one of which is the fact that you must not persist anything at this stage of the transition logic. If you persist something and after than on `token.update` one of the validations fails, we are not going to be able to revert the persisted changes